tests/attr/test_data_parallel.py

```diff
@@ -1,898 +1,226 @@
-            )
-
-    def _alt_device_list(self):
-        return [0] + [x for x in range(torch.cuda.device_count() - 1, 0, -1)]
-
-    def _data_parallel_test_assert(
-        self,
-        algorithm,
-        model,
-        target_layer=None,
-        alt_device_ids=False,
-        test_batches=False,
-        delta=0.0001,
-        **kwargs
-    ):
-        if alt_device_ids:
-            dp_model = torch.nn.parallel.DataParallel(
-                model, device_ids=self._alt_device_list()
-            )
-        else:
-            dp_model = torch.nn.parallel.DataParallel(model)
-        if target_layer:
-            attr_orig = algorithm(model, target_layer)
-            if alt_device_ids:
-                attr_dp = algorithm(
-                    dp_model.forward, target_layer, device_ids=self._alt_device_list()
-                )
-            else:
-                attr_dp = algorithm(dp_model, target_layer)
-        else:
-            attr_orig = algorithm(model)
-            attr_dp = algorithm(dp_model)
-
-        batch_sizes = [None]
-        delta_orig = None
-        delta_dp = None
-        if test_batches:
-            batch_sizes = [None, 1, 8]
-        for batch_size in batch_sizes:
-            if batch_size:
-                attributions_orig = attr_orig.attribute(
-                    internal_batch_size=batch_size, **kwargs
-                )
-            else:
-                if attr_orig.has_convergence_delta():
-                    attributions_orig, delta_orig = attr_orig.attribute(
-                        return_convergence_delta=True, **kwargs
+            for algorithm in algorithms:
+                for mode in DataParallelCompareMode:
+                    # Creates test case corresponding to each algorithm and
+                    # DataParallelCompareMode
+                    test_method = cls.make_single_dp_test(
+                        algorithm,
+                        model,
+                        target_layer,
+                        args,
+                        dp_delta,
+                        noise_tunnel,
+                        baseline_distr,
+                        mode,
+                    )
+                    test_name = (
+                        "test_"
+                        + cast(str, test_config["name"])
+                        + "_"
+                        + mode.name
+                        + "_"
+                        + algorithm.__name__
+                        + ("NoiseTunnel" if noise_tunnel else "")
+                    )
+                    if test_name in attrs:
+                        raise AssertionError(
+                            "Trying to overwrite existing test with name: %r"
+                            % test_name
+                        )
+                    attrs[test_name] = test_method
+
+        return super(DataParallelMeta, cls).__new__(cls, name, bases, attrs)
+
+    @classmethod
+    @deep_copy_args
```

Comment:
This is primarily needed for test_targets, since we modify the args each time we run each example independently. We can reset it at the end, but if the test errors, it leaves the modified args for the next test. I thought it's safer to deep copy all arguments to ensure that all tests are independent and not dependent on any modifications made in a previous test, so I included this in all methods. Will add a short comment explaining this.